gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,432 @@
+            dim=1,
+        )
+        self.regularizer_arguments["coords"] = self.regularizer_arguments[
+            "regularization_dataset"
+        ]
+        self.regularizer_arguments["connections"] = connections
+        self.regularizer_arguments["dummy_loader"] = dummy_loader
+        return True
+
+    def update_params(self, regularizer_arguments):
+        """
+        Finds the critical vertices of the linear surrogate function
+        """
+        network = self.regularizer_arguments["model"]
+        coords = self.regularizer_arguments["coords"]
+        dummy_loader = self.regularizer_arguments["dummy_loader"]
+        connections = self.regularizer_arguments["connections"]
+        expansion = self.regularizer_arguments["expansion"]
+        cpoints1 = compute_critical_points(
+            network, coords, dummy_loader, connections, expansion, ind=0
+        )
+        cpoints2 = compute_critical_points(
+            network, coords, dummy_loader, connections, expansion, ind=1
+        )
+        cpoints = np.concatenate(
+            [cpoints1.detach().numpy(), cpoints2.detach().numpy()], 0
+        )
+        cpoints = torch.tensor(cpoints)
+        cpoints.requires_grad = True
+        return cpoints
+        # pass
+
+    def regularization_penalty(self, forward_passed_arguments, model: torch.nn.Module):
+        """
+        The penalty is the squared sum of
+        the function values at the critical points.
+        """
+        values = model(forward_passed_arguments)
+        tmp = torch.max(F.softmax(values, 1) - 0.5, 1)
+        return torch.sum((tmp.values) ** 2)
+
+
+class TihonovRegularizer(Regularizer):
+    """
+    An example implementation of a p-norm regularizer
+    This regularizer penalizes the p-norm (to the pth power) of the network weights.
+    For example, p=1 is LASSO and p=2 is the Ridge Regression.
+    reg_params: p (int), (Optional)
+    """
+
+    def preprocess(self, **regularizer_arguments: Any):
+        defaultKwargs = {"p": 2}
+        self.params = {**defaultKwargs, **regularizer_arguments}
+        """
+        Tihonov-type regularizer needs no preprocessing
+        """
+        return True
+
+    def update_params(self, regularizer_arguments: Any):
+        """
+        Tihonov regularizer needs no non-differentiable updates
+        """
+        return regularizer_arguments
+
+    def regularization_penalty(
+        self, forward_passed_arguments: Any, model: torch.nn.Module
+    ) -> Tensor:
+        """
+        The regularization penalty is sum of absolute values of the
+        regression coefficients, i.e. network weights
+        """
+        total = torch.tensor(0)  # , dtype=float)
+        for parameter in model.parameters():
+            total = total + torch.norm(parameter, self.params["p"]) ** self.params["p"]
+        return total
+
+
+class RegularizedTrainer(Trainer):
+    def __init__(self, regularizer, reg_params={}, *args, **kwargs):
```

Comment:
Add type hint to the parameter `regularizer`.

Reply:
done, (class removed)